File: scala-project/src/main/scala/com/waitingforcode/DevicesToDeltaConverter.scala
package com.waitingforcode

import org.apache.spark.sql.{SaveMode, SparkSession}

object DevicesToDeltaConverter {

  def convertDevicesToDeltaTable(sparkSession: SparkSession, devicesInput: String,
                                 devicesOutput: String): Unit = {
    // The data generator doesn't support Delta Lake sink yet; converting it quickly before starting the job
    val inputDevices = sparkSession.read.schema("type STRING, full_name STRING, version STRING").json(devicesInput)

    inputDevices.write.mode(SaveMode.Overwrite).format("delta").save(devicesOutput)
  }

}

File: python-project/visits_generator/devices_to_delta_converter.py
from pyspark.sql import SparkSession


def convert_devices_to_delta_table(spark_session: SparkSession, devices_input_location: str,
                                   devices_output_location: str):
    input_devices = spark_session.read.schema('type STRING, full_name STRING, version STRING').json(devices_input_location)

    input_devices.write.mode('overwrite').format('delta').save(devices_output_location)
